File: LLD/executors_and_callables/ImagePainter/ImagePainter.py
# Parallel Matrix Painting
# Problem Statement
# You are tasked with creating a Python program that can divide a matrix into quadrants and paint each quadrant with a different color using parallel processing. The program should leverage the concurrent.futures module to achieve parallelism and improve performance.

# Requirements
# The program should include the following components:

# paint_quadrant function:
# Takes a matrix, start and end row indices, start and end column indices, and a color as input.
# Paints the specified quadrant of the matrix with the given color.
# divideandpaint function:
# Takes a matrix and a list of colors as input.
# Divides the matrix into four quadrants (top-left, top-right, bottom-left, bottom-right).
# Utilizes the concurrent.futures.ThreadPoolExecutor to create four separate threads.
# Each thread calls the paint_quadrant function with the corresponding quadrant and color.
# Waits for all threads to complete before returning.
# Example usage:
# Creates a sample 4x4 matrix.
# Calls the divideandpaint function with the matrix and a list of four colors.
# Prints the painted matrix.
# Instructions
# Implement the paint_quadrant function according to the specified requirements.
# Implement the divideandpaint function, which creates a thread pool executor and submits tasks to paint each quadrant in parallel.

import concurrent.futures

def paint_quadrant(matrix, start_row, end_row, start_col, end_col, color):
    """Paints a specific quadrant of the matrix with the given color."""
    for i in range(start_row, end_row):
        for j in range(start_col, end_col):
            matrix[i][j] = color

def divide_and_paint(matrix, colors):
    """Divides the matrix into four quadrants and paints them in parallel."""
    rows, cols = len(matrix), len(matrix[0])
    mid_row, mid_col = rows // 2, cols // 2  # Find midpoints to divide into quadrants

    # Define the quadrants
    quadrants = [
        (0, mid_row, 0, mid_col, colors[0]),  # Top-left
        (0, mid_row, mid_col, cols, colors[1]),  # Top-right
        (mid_row, rows, 0, mid_col, colors[2]),  # Bottom-left
        (mid_row, rows, mid_col, cols, colors[3])  # Bottom-right
    ]

    # Use ThreadPoolExecutor for parallel execution
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(paint_quadrant, matrix, *q) for q in quadrants]
        
        # Wait for all tasks to complete
        concurrent.futures.wait(futures)

# Example usage:
matrix = [[0, 0, 0, 0],
          [0, 0, 0, 0],
          [0, 0, 0, 0],
          [0, 0, 0, 0]]

# Colors assigned to quadrants
colors = ["red", "blue", "green", "yellow"]

# Apply the painting function
divide_and_paint(matrix, colors)

# Print the painted matrix
for row in matrix:
    print(row)

File: LLD/executors_and_callables/ImagePainter/ImagePainter.java
package LLD.executors_and_callables.ImagePainter;
/*
 * Image Painter
Implement a Java program that simulates image re-painting using 4 threads. The algorithm for painting is described as under but your task is given an image, divide the image into 4 quadrants and let each thread paint one quadrant of the image. The concurrent execution ensures image is painted efficiently.

Problem Statement:
You are provided a function paintImage. This function is designed to paint a square image, represented as a 2D integer array, where each integer represents a color. The goal is to paint the image in such a way that each quadrant of the image has a different color. The painting process for each quadrant should be performed concurrently using multiple threads.

Tasks:
Implement the paintImage function
Divide the given image into four quadrants.
Paint each quadrant with a different color.
The base color will be provided as an input to the painting method, and each quadrant should use a variation of this base color.
For first quadrant the color should be same as the given base color.
For second quadrant the color should be equal to 2*base color.
For third quadrant the color should be equal to 3*base color.
For fourth quadrant the color should be equal to 4*base color
Instructions
The image is represented as a 2D integer array.
Use a fixed thread pool size of 4 threads in your ExecutorService.
Handle any potential exceptions that may arise during the execution.
Implement the paintImage function and the constructor in the ImagePainter class.
Use Java's ExecutorService to execute the painting of each quadrant in a separate thread.
Ensure that the program waits for all quadrants to be fully painted before proceeding.
Properly shutdown your ExecutorService after the painting is completed to avoid any resource leaks.
 */

 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.ExecutionException;
 
 public class ImagePainter {
     private final ExecutorService executor;  // Instance-level executor
     private final int[][] image;
 
     // Initialize the image and create an executor with a fixed thread pool size of 4
     public ImagePainter(int[][] image) {
         this.image = image;
         this.executor = Executors.newFixedThreadPool(4);  // Create a new executor for each instance
     }
 
     public void paintImage(int color) throws InterruptedException {
         int rows = image.length;
         int cols = image[0].length;
         int halfRows = rows / 2;
         int halfCols = cols / 2;
 
         // Submitting tasks for each quadrant
         Future<?> q1 = executor.submit(() -> paintQuadrant(0, halfRows, 0, halfCols, color));      // Top-left
         Future<?> q2 = executor.submit(() -> paintQuadrant(0, halfRows, halfCols, cols, 2 * color)); // Top-right
         Future<?> q3 = executor.submit(() -> paintQuadrant(halfRows, rows, 0, halfCols, 3 * color)); // Bottom-left
         Future<?> q4 = executor.submit(() -> paintQuadrant(halfRows, rows, halfCols, cols, 4 * color)); // Bottom-right
 
         try {
             q1.get();
             q2.get();
             q3.get();
             q4.get();
         } catch (ExecutionException e) {
             e.printStackTrace();
         }
 
         // Shutdown the executor after completing all tasks
         executor.shutdown();
     }
 
     private void paintQuadrant(int startRow, int endRow, int startCol, int endCol, int color) {
         for (int i = startRow; i < endRow; i++) {
             for (int j = startCol; j < endCol; j++) {
                 image[i][j] = color;
             }
         }
     }
 
     public int[][] getImage() {
         return image;
     }
 }
